modules/edge-ren/src/tests.rs

```diff
@@ -0,0 +1,787 @@
+//! Unit tests for the renvm bridge module.
+
+
+#![cfg(test)]
+
+use super::*;
+use frame_support::{assert_err, assert_noop, assert_ok, unsigned::ValidateUnsigned};
+use hex_literal::hex;
+use mock::{EdgeAssets, AccountId, ExtBuilder, Origin, RenVmBridge, RenvmBridgeCall, System};
+use sp_runtime::transaction_validity::TransactionValidityError;
+use sp_runtime::{DispatchError,traits::Dispatchable};
+
+
+fn mint_ren_token(
+	_ren_token_id: u32,
+	who: AccountId,
+	p_hash: [u8; 32],
+	amount: mock::Balance,
+	n_hash: [u8; 32],
+	sig: EcdsaSignature,
+) -> Result<DispatchResult, TransactionValidityError> {
+	<RenVmBridge as ValidateUnsigned>::validate_unsigned(
+		TransactionSource::External,
+		&RenvmBridgeCall::mint(_ren_token_id, who.clone(), p_hash, amount, n_hash, sig.clone()),
+	)?;
+
+	Ok(RenVmBridge::mint(Origin::none(), _ren_token_id, who, p_hash, amount, n_hash, sig))
+}
+
+
+#[test]
+fn token_mint_fails_on_bad_init_but_works_after() {
+	ExtBuilder::default().build().execute_with(|| {
+
+		assert_ok!(mock::Call::EdgeAssets(edge_assets::Call::force_create(
+				0,
```

Comment:
Indentation on these tests is off. Please make sure it matches other repos with 4 spaced tabs.